src/Aspire.Hosting/Eventing/IDistributedApplicationEventing.cs

```diff
@@ -0,0 +1,30 @@
+namespace Aspire.Hosting.Eventing;
+
+/// <summary>
+/// Supports publishing and subscribing to events which are executed during the AppHost lifecycle.
+/// </summary>
+public interface IDistributedApplicationEventing
+{
+    /// <summary>
+    /// Subscribes a callback to a specific event type within the AppHost.
+    /// </summary>
+    /// <typeparam name="T">The type of the event.</typeparam>
+    /// <param name="callback">A callback to handle the event.</param>
+    /// <returns>A subscription instance which can be used to unsubscribe </returns>
+    DistributedApplicationEventSubscription Subscribe<T>(Func<T, CancellationToken, Task> callback) where T : IDistributedApplicationEvent;
+
+    /// <summary>
+    /// Unsubscribe from an event.
+    /// </summary>
+    /// <param name="subscription">The specific subscription to unsubscribe.</param>
+    void Unsubscribe(DistributedApplicationEventSubscription subscription);
```

Comment:
Make the DistributedApplicationEventSubscription, IDisposable instead?

Reply:
Even if we make it disposable we need the unsubscribe :)